libbpf-rs/src/btf.rs

```diff
@@ -0,0 +1,440 @@
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// The number of [BtfType]s in this object.
+    pub fn len(&self) -> usize {
+        (unsafe { libbpf_sys::btf__type_cnt(self.ptr.as_ptr()) }) as usize
+    }
+
+    /// Find a btf type by name
+    ///
+    /// # Panics
+    /// If `name` has null bytes.
+    pub fn type_by_name<'s, K>(&'s self, name: &str) -> Option<K>
+    where
+        K: TryFrom<BtfType<'btf, 's>>,
+        'btf: 's,
+    {
+        let ty = unsafe {
+            let c_string = CString::new(name)
+                .map_err(|_| Error::InvalidInput(format!("{name:?} contains null bytes")))
+                .unwrap();
+            libbpf_sys::btf__find_by_name(self.ptr.as_ptr(), c_string.as_ptr())
+        };
+        if ty < 0 {
+            None
+        } else {
+            self.type_by_id(TypeId(ty as _))
+        }
+    }
+
+    /// Find a type by it's [TypeId].
+    pub fn type_by_id<'s, K>(&'s self, type_id: TypeId) -> Option<K>
+    where
+        K: TryFrom<BtfType<'btf, 's>>,
+        'btf: 's,
+    {
+        let btf_type = unsafe { libbpf_sys::btf__type_by_id(self.ptr.as_ptr(), type_id.0) };
+
+        if btf_type.is_null() {
+            return None;
+        }
+        let ty = unsafe { &*btf_type };
+        let name = self.name_at(ty.name_off);
+        BtfType {
+            type_id,
+            name,
+            source: self,
+            ty,
+        }
+        .try_into()
+        .ok()
+    }
+
+    /// Find all type of a specific type kind.
+    pub fn type_by_kind<'s, K>(&'s self) -> TypeByKindIter<'btf, 's, K>
+    where
+        K: TryFrom<BtfType<'btf, 's>>,
+        'btf: 's,
+    {
+        TypeByKindIter {
+            len: self.len() as u32,
+            index: 1,
+            btf: self,
+            _marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Iterator produced by [`Btf::type_by_kind`]
+pub struct TypeByKindIter<'btf, 's, K> {
+    btf: &'s Btf<'btf>,
+    index: u32,
+    len: u32,
+    _marker: PhantomData<K>,
+}
+
+impl<'btf, 's, K> Iterator for TypeByKindIter<'btf, 's, K>
+where
+    K: TryFrom<BtfType<'btf, 's>>,
+    'btf: 's,
+{
+    type Item = K;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (id, k) = (self.index..self.len).map(TypeId::from).find_map(|id| {
+            let ty = self.btf.type_by_id(id)?;
+            let k = K::try_from(ty).ok()?;
+            Some((id, k))
+        })?;
+
+        self.index = id.0 + 1;
+        Some(k)
+    }
+}
+
+impl Drop for Btf<'_> {
+    fn drop(&mut self) {
+        if self.needs_drop {
+            unsafe { libbpf_sys::btf__free(self.ptr.as_ptr()) }
+        }
+    }
+}
+
+/// A undiscriminated btf type
+///
+/// For the documentation of the lifetime `'btf` see [`Btf`].
+///
+/// The lifetime `'s` guards against this type outliving the [`Btf`] instance it came from.
+#[derive(Clone, Copy)]
+pub struct BtfType<'btf, 's>
+where
+    'btf: 's,
```

Comment:
You're right, I was trying to be smart here because the lifetime of the name is sometimes longer than the lifetime of the Btf, in the case where the name comes from a `bpf_object`, but sadly without going full on type magic, I couldn't make it work (hence the comment on the `name` field).

What I failed to notice was that I could trim the lifetime after that attempt failed and indeed I can make this simpler.